libraries/chain/webassembly/eos-vm-oc/compile_monitor.cpp

```diff
@@ -0,0 +1,309 @@
+            [&](const auto&) {
+               //anything else is an error
+            }
+         });
+
+         wait_for_new_incomming_session(ctx);
+      });
+   }
+
+   local::datagram_protocol::socket _nodeos_socket;
+   wrapped_fd _trampoline_socket;
+
+   std::list<compile_monitor_session> _compile_sessions;
+};
+
+void launch_compile_monitor(int nodeos_fd) {
+   prctl(PR_SET_NAME, "oc-monitor");
+   prctl(PR_SET_PDEATHSIG, SIGKILL);
+
+   //first off, let's disable shutdown signals to us; we want all shutdown indicators to come from
+   // nodeos shutting us down
+   sigset_t set;
+   sigemptyset(&set);
+   sigaddset(&set, SIGHUP);
+   sigaddset(&set, SIGTERM);
+   sigaddset(&set, SIGPIPE);
+   sigaddset(&set, SIGINT);
+   sigaddset(&set, SIGQUIT);
+   sigprocmask(SIG_BLOCK, &set, nullptr);
+
+   int socks[2]; //0: local trampoline socket, 1: the one we give to trampoline
+   socketpair(AF_UNIX, SOCK_SEQPACKET | SOCK_CLOEXEC, 0, socks);
+   pid_t child = fork();
+   if(child == 0) {
+      close(socks[0]);
+      run_compile_trampoline(socks[1]);
+   }
+   close(socks[1]);
+
+   {
+      boost::asio::io_context ctx;
+      boost::asio::local::datagram_protocol::socket nodeos_socket(ctx);
+      nodeos_socket.assign(boost::asio::local::datagram_protocol(), nodeos_fd);
+      wrapped_fd trampoline_socket(socks[0]);
+      compile_monitor monitor(ctx, std::move(nodeos_socket), std::move(trampoline_socket));
+      ctx.run();
+      if(monitor._compile_sessions.size())
+         std::cerr << "ERROR: EOS-VM OC compiler monitor exiting with active sessions" << std::endl;
+   }
+   
+   _exit(0);
+}
+
+struct compile_monitor_trampoline {
+   void start() {
+      //create communication socket; let's hold off on asio usage until all forks are done
+      int socks[2];
+      socketpair(AF_UNIX, SOCK_SEQPACKET | SOCK_CLOEXEC, 0, socks);
+      compile_manager_fd = socks[0];
+
+      pid_t child = fork();
+      if(child == 0) {
+         close(socks[0]);
+         launch_compile_monitor(socks[1]);
+      }
+      close(socks[1]);
+   }
+
+   pid_t compile_manager_pid = -1;
+   int compile_manager_fd = -1;
+};
+
+static compile_monitor_trampoline the_compile_monitor_trampoline;
+extern "C" int __real_main(int, char*[]);
+extern "C" int __wrap_main(int argc, char* argv[]) {
+   the_compile_monitor_trampoline.start();
+   return __real_main(argc, argv);
+}
+
+wrapped_fd get_connection_to_compile_monitor(int cache_fd) {
+   FC_ASSERT(the_compile_monitor_trampoline.compile_manager_fd >= 0, "EOS-VM oop connection doesn't look active");
+
+   int socks[2]; //0: our socket to compile_manager_session, 1: socket we'll give to compile_maanger_session
+   socketpair(AF_UNIX, SOCK_SEQPACKET | SOCK_CLOEXEC, 0, socks);
+   wrapped_fd socket_to_monitor_session(socks[0]);
+
+   std::vector<wrapped_fd> fds_to_pass; 
+   fds_to_pass.emplace_back(socks[1]);
+   fds_to_pass.emplace_back(cache_fd);  //put cache_fd in to a wrapped_fd just temporarily, ick
```

Comment:
This is especially bad, because it will close cache_fd if write_message_with_fds throws, but not if any other part of the function throws.